clang/lib/Headers/hlsl/hlsl_intrinsics.h

```diff
@@ -387,6 +387,65 @@ float3 asin(float3);
 _HLSL_BUILTIN_ALIAS(__builtin_elementwise_asin)
 float4 asin(float4);
 
+//===----------------------------------------------------------------------===//
+// asuint builtins
+//===----------------------------------------------------------------------===//
+
+/// \fn uint asuint(T Val)
+/// \brief Interprets the bit pattern of x as an unsigned integer.
+/// \param Val The input value.
+
+__attribute__((__always_inline__, __nodebug__)) static inline uint
```

Comment:
I don't really like how dissimilar this looks to all the other intrinsics. Maybe we could capture these new attributes in a new preprocessor macro. Something like `_HLSL_CAST_BUILTIN_ALIAS`?

Reply:
@farzonl Let me know if you still think this is necessary after changing the implementation to use templates.

Reply:
I kind of disagree with this, and I find the version with the macro pretty hard to follow.
1. Even if we have a macro to help with the attributes or something, it shouldn't be called "alias" - this isn't an alias at all (because of the static assert).
2. I don't really like that this single macro is defining multiple functions. The uses of `_HLSL_BUILTIN_ALIAS` decorate the definitions but each definition is written out clearly in source.

I could see some value in putting the `__attribute__((__always_inline__, __nodebug__))` bit in a macro so that it's easy to be consistent, but I don't think the function definitions themselves should be done this way.

Reply:
Justin is right, the static assert makes this no longer just an alias for `__builtin_bit_cast` so the macro name has to change at a minimum. 

Also we mostly used macros to cover our attribute usage. So this is as much a new pattern as what you had. That said, while this isn't what I imagined this to look like I'm not opposed to how this turned out if the function body you defined is going to be reused across all our cast apis? That all depends on if the remaining cast work is boilerplate enough that this is flexible enough to support it.

Reply:
I think there might be some things we are unaware of right now. I think we can use the macros to host the attributes, or help us do some of the `static_cast`.  In the future, once we've implemented more of the casting functionality, we can reassess and see if there is more boilerplate that can be moved into the macro.